src/lib/routes/admin-api/user-admin.ts

```diff
@@ -0,0 +1,144 @@
+import Controller from '../controller';
+import { ADMIN } from '../../permissions';
+import { IUnleashConfig } from '../../types/core';
+import UserService from '../../services/user-service';
+import { AccessService, RoleName } from '../../services/access-service';
+import { Logger } from '../../logger';
+
+class UserAdminController extends Controller {
+    private userService: UserService;
+
+    private accessService: AccessService;
+
+    private logger: Logger;
+
+    constructor(config: IUnleashConfig, { userService, accessService }) {
+        super(config);
+        this.userService = userService;
+        this.accessService = accessService;
+        this.logger = config.getLogger('routes/user-controller.js');
+
+        this.get('/', this.getUsers);
+        this.get('/search', this.search);
+        this.post('/', this.createUser, ADMIN);
+        this.post('/validate-password', this.validatePassword);
+        this.put('/:id', this.updateUser, ADMIN);
+        this.post('/:id/change-password', this.changePassword, ADMIN);
+        this.delete('/:id', this.deleteUser, ADMIN);
+    }
+
+    private validateRootRole(roleName) {
+        if (!Object.values(RoleName).includes(roleName)) {
+            throw new Error(`Unknown rootRole = ${roleName}`);
+        }
+    }
+
+    async getUsers(req, res) {
+        try {
+            const users = await this.userService.getAll();
+            const rootRoles = await this.accessService.getRootRolesForAllUsers();
+
+            const usersWithRootRole = users.map(u => {
+                const rootRole = rootRoles.find(r => r.userId === u.id);
+                const rootRoleName =
+                    rootRole && rootRole.roles.length > 0
+                        ? rootRole.roles[0].name
+                        : RoleName.READ;
+                return { ...u, rootRole: rootRoleName };
+            });
+
+            res.json({ users: usersWithRootRole });
+        } catch (error) {
+            this.logger.error(error);
+            res.status(500).send({ msg: 'server errors' });
+        }
+    }
+
+    async search(req, res) {
+        const { q } = req.query;
+        try {
+            const users =
+                q && q.length > 1 ? await this.userService.search(q) : [];
+            res.json(users);
+        } catch (error) {
+            this.logger.error(error);
+            res.status(500).send({ msg: 'server errors' });
+        }
+    }
+
+    async createUser(req, res) {
+        const { username, email, name, rootRole = 'Regular' } = req.body;
```

Comment:
root has a very specific connotation for access control in my unix/linux head so saying `rootRole = 'Regular'` sounds unfamiliar to me. It might just be me, but to me, root is a very specific role with full access to make all changes. Would `baseRole` fulfill the same or have you allocated a meaning to rootRole that I miss because I'm assuming default unix model?

Reply:
It will require some refactoring and new db-migrations as we already landed this in the [RBAC specification](https://docs.google.com/document/d/1T2iuM9qgbgPfNMqXoQKEgdEG--a93lFzMsJHHuT0MHI/edit)

Reply:
That's what I was afraid of, and shows the danger of too large PRs, it escaped me when I read the RBAC spec. We can own the new spec, it shouldn't require too many takes before it becomes natural :)